rasterio/_io.pyx

```diff
@@ -306,17 +319,23 @@ cdef class DatasetReaderBase(DatasetBase):
 
         else:
             # Compute the overlap between the dataset and the boundless window.
-            overlap = ((
-                max(min(window[0][0], self.height), 0),
-                max(min(window[0][1], self.height), 0)), (
-                max(min(window[1][0], self.width), 0),
-                max(min(window[1][1], self.width), 0)))
-
-            if overlap != ((0, 0), (0, 0)):
+            try:
+                overlap = intersection(
+                    window, Window(0, 0, self.width, self.height))
+            except WindowError:
+                log.info("Dataset and window do not overlap")
+                data = None
+                if masked:
+                    kwds = {'mask': True}
+                    if len(set(nodatavals)) == 1:
+                        if nodatavals[0] is not None:
+                            kwds['fill_value'] = nodatavals[0]
+                    out = np.ma.array(out, **kwds)
+            else:
```

Comment:
I like it. Using exceptions for control flow is Python to the core :-) Raymond Hettinger wrote a solid stackoverflow response on the topic: https://stackoverflow.com/a/16138864

I really like having the post-conditions (code run after success or hitting an Exception) at the same level of indentation. Sorta silly but readability counts

```
try:
    something_that_might_fail()
except SomeError:
    fail_gracefully()
else:
    do_next_step()
```